src/MySqlConnector/Core/ServerSession.cs

```diff
@@ -225,60 +225,109 @@ public void FinishQuerying()
+					m_supportsConnectionAttributes = (initialHandshake.ProtocolCapabilities & ProtocolCapabilities.ConnectionAttributes) != 0;
+					m_supportsDeprecateEof = (initialHandshake.ProtocolCapabilities & ProtocolCapabilities.DeprecateEof) != 0;
+					var serverSupportsSsl = (initialHandshake.ProtocolCapabilities & ProtocolCapabilities.Ssl) != 0;
 
-				Log.Info("{0} made connection; ServerVersion={1}; ConnectionId={2}; Flags: {3}{4}{5}{6}", m_logArguments[0], ServerVersion.OriginalString, ConnectionId,
-					m_useCompression ? "Cmp " :"", m_supportsConnectionAttributes ? "Attr " : "", m_supportsDeprecateEof ? "" : "Eof ", serverSupportsSsl ? "Ssl " : "");
+					Log.Info("{0} made connection; ServerVersion={1}; ConnectionId={2}; Flags: {3}{4}{5}{6}", m_logArguments[0], ServerVersion.OriginalString, ConnectionId,
+						m_useCompression ? "Cmp " : "", m_supportsConnectionAttributes ? "Attr " : "", m_supportsDeprecateEof ? "" : "Eof ", serverSupportsSsl ? "Ssl " : "");
 
-				if (cs.SslMode != MySqlSslMode.None && (cs.SslMode != MySqlSslMode.Preferred || serverSupportsSsl))
-				{
-					if (!serverSupportsSsl)
+					if (cs.SslMode != MySqlSslMode.None && (cs.SslMode != MySqlSslMode.Preferred || serverSupportsSsl))
+					{
+						if (!serverSupportsSsl)
+						{
+							Log.Error("{0} requires SSL but server doesn't support it", m_logArguments);
+							throw new MySqlException("Server does not support SSL");
+						}
+
+						// MySQL Server 8.0 and MariaDB are compiled with OpenSSL, which negotiates TLS 1.2 correctly; a failure should
+						// not cause an automatic retry with an older/weaker TLS version.
+						var serverHasOpenSsl = shouldRetrySsl && (ServerVersion.Version.Major >= 8 || ServerVersion.OriginalString.IndexOf("MariaDB", StringComparison.OrdinalIgnoreCase) != -1);
+
+						try
+						{
+							await InitSslAsync(initialHandshake.ProtocolCapabilities, cs, sslProtocols, ioBehavior, cancellationToken).ConfigureAwait(false);
+							shouldRetrySsl = false;
+						}
+#if NET45 || NET46
+						catch (ArgumentException ex) when (shouldRetrySsl && ex.ParamName == "sslProtocolType")
+						{
+							// Prior to .NET Framework 4.7, SslProtocols.None is not a valid argument to AuthenticateAsClientAsync.
+							// If the NET46 build is loaded by an application that targets. NET 4.7, SslProtocols.None will work; however,
+							// if the application targets .NET 4.6.2 or earlier, then it will fail at runtime; we catch the exception and retry.
+							Log.Warn(ex, "{0} doesn't support SslProtocols.None; retrying with SslProtocols.Tls12", m_logArguments);
+							sslProtocols = SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12;
+							if (Pool != null)
+								Pool.SslProtocols = sslProtocols;
+						}
+#endif
+						catch (Exception ex) when (shouldRetrySsl && serverHasOpenSsl && ((ex is MySqlException && ex.InnerException is IOException) || ex is IOException))
```

Comment:
This logic is reversed; it was supposed to be `!serverHasOpenSsl`.

This code fails when tested against a local instance of MySQL Server 5.7.21 for Windows (but apparently not on Appveyor?).